fs.go

```diff
@@ -0,0 +1,197 @@
+package dalec
+
+import (
+	"context"
+	"io"
+	"io/fs"
+	"path"
+	"sync"
+
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/tonistiigi/fsutil/types"
+
+	"github.com/tonistiigi/fsutil"
+
+	"github.com/moby/buildkit/client/llb"
+)
+
+var _ fs.DirEntry = &stateRefDirEntry{}
+var _ fs.ReadDirFS = &StateRefFS{}
+
+type StateRefFS struct {
+	s       llb.State
+	ctx     context.Context
+	opts    []llb.ConstraintsOpt
+	client  gwclient.Client
+	res     *gwclient.Result
+	o       sync.Once
+	initErr error
+}
+
+func NewStateRefFS(s llb.State, ctx context.Context, client gwclient.Client) *StateRefFS {
+	return &StateRefFS{
+		s:      s,
+		ctx:    ctx,
+		client: client,
+	}
+}
+
+func (fs *StateRefFS) fetchRef() (*gwclient.Result, error) {
+	def, err := fs.s.Marshal(fs.ctx, fs.opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := fs.client.Solve(fs.ctx, gwclient.SolveRequest{
+		Definition: def.ToPB(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (fs *StateRefFS) Res() (*gwclient.Result, error) {
+	fs.o.Do(func() {
+		res, err := fs.fetchRef()
+		fs.res = res
+		fs.initErr = err
+	})
+
+	return fs.res, fs.initErr
+}
+
+func (fs *StateRefFS) Ref() (gwclient.Reference, error) {
+	res, err := fs.Res()
+	if err != nil {
+		return nil, err
+	}
+
+	ref, err := res.SingleRef()
+	if err != nil {
+		return nil, err
+	}
+
+	return ref, nil
+}
+
+type stateRefDirEntry struct {
+	stat *types.Stat
+}
+
+func (s *stateRefDirEntry) Name() string {
+	return path.Base(s.stat.Path)
+}
+
+func (s *stateRefDirEntry) IsDir() bool {
+	return s.stat.IsDir()
+}
+
+func (s *stateRefDirEntry) Type() fs.FileMode {
+	return fs.FileMode(s.stat.Mode)
+}
+
+func (s *stateRefDirEntry) Info() (fs.FileInfo, error) {
+	info := &fsutil.StatInfo{
+		Stat: s.stat,
+	}
+
+	return info, nil
+}
+
+func (st *StateRefFS) ReadDir(name string) ([]fs.DirEntry, error) {
+	ref, err := st.Ref()
+	if err != nil {
+		return nil, err
+	}
+
+	contents, err := ref.ReadDir(st.ctx, gwclient.ReadDirRequest{
+		Path: name,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	entries := []fs.DirEntry{}
+	for _, stat := range contents {
+		dirEntry := &stateRefDirEntry{stat: stat}
+		entries = append(entries, dirEntry)
+	}
+
+	return entries, nil
+}
+
+type stateRefFile struct {
+	eof    bool   // has file been read to EOF?
+	path   string // the full path of the file from root
+	ref    gwclient.Reference
+	ctx    context.Context
+	stat   *types.Stat
+	offset int64
+}
+
+// close is a no-op
+func (s *stateRefFile) Close() error {
+	return nil
+}
+
+func (s *stateRefFile) Read(b []byte) (int, error) {
```

Comment:
`stateRefFile` should be able to implement `io.ReaderAt`, and then `Read` can be implemented on top of that.

Reply:
This is done!